### PR TITLE
IGNITE-16943: Implement Micronaut REST

### PR DESCRIPTION
**Pull request by PakhomovAlexander**
- Introduce micronaut-based rest server
- Generate openapi.yaml file from server implementation
- Add rest-client module that produces auto-generated rest client
- Use micronaut-serde library in order to avoid jackson dependency

### CODE REVIEW

File: parent/pom.xml [hunk 1]
```diff
@@ -53,19 +53,21 @@
...
         <netty.version>4.1.70.Final</netty.version>
         <javapoet.version>1.13.0</javapoet.version>
         <javax.annotation.api.version>1.3.2</javax.annotation.api.version>
         <jetbrains.annotations.version>20.1.0</jetbrains.annotations.version>
         <jmh.framework.version>1.35</jmh.framework.version>
         <junit.version>5.8.1</junit.version>
-        <micronaut.version>2.1.2</micronaut.version>
-        <micronaut.test.junit5.version>2.3.1</micronaut.test.junit5.version>
+        <micronaut.version>3.4.1</micronaut.version>
+        <micronaut.test.junit5.version>3.0.5</micronaut.test.junit5.version>
```
Comment (sashapolo) [inline parent/pom.xml:65]: I think we can get rid of this property if we use a bom

File: parent/pom.xml [hunk 2]
```diff
@@ -954,6 +963,7 @@
                             <ignorePackages>
                                 <package>org.apache.ignite.internal.sql.engine</package>
                                 <package>org.apache.ignite.internal.generated</package>
+                                <package>org.apache.ignite.rest.client</package>
```
Comment (sashapolo) [inline parent/pom.xml:966]: Do I understand correctly that this package contains clients generated by Swagger? If yes, please add a comment, if no - why did you add it here?

File: modules/rest-client/pom.xml [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    <version>1</version>
+    <relativePath>../../parent/pom.xml</relativePath>
+  </parent>
+
+  <artifactId>ignite-rest-client</artifactId>
+  <version>3.0.0-SNAPSHOT</version>
+
+  <dependencies>
+
+    <!--client specific dependencies, versions cannot be moved to the parent pom-->
```
Comment (sashapolo) [inline modules/rest-client/pom.xml:37]: Is this module (and this file) auto-generated?

File: modules/cluster-management/src/integrationTest/java/org/apache/ignite/internal/cluster/management/ItClusterManagerTest.java [hunk 4]
```diff
@@ -107,13 +107,13 @@ void testInit(TestInfo testInfo) throws Exception {
 
         initCluster(metaStorageNodes, cmgNodes);
 
-        assertThat(cluster.get(0).clusterManager().metaStorageNodes(), will(containsInAnyOrder(metaStorageNodes)));
-        assertThat(cluster.get(1).clusterManager().metaStorageNodes(), will(containsInAnyOrder(metaStorageNodes)));
+        assertThat(cluster.get(0).getClusterManager().metaStorageNodes(), will(containsInAnyOrder(metaStorageNodes)));
+        assertThat(cluster.get(1).getClusterManager().metaStorageNodes(), will(containsInAnyOrder(metaStorageNodes)));
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/cluster-management/src/integrationTest/java/org/apache/ignite/internal/cluster/management/ItClusterManagerTest.java:111]: Thanks for the note, I think this was a mistake. I'll revert this back.

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/config/ConfigurationClient.java [hunk 5]
```diff
@@ -81,7 +81,7 @@ public String get(
     ) {
         var req = HttpRequest
                 .newBuilder()
-                .header("Content-Type", "application/json");
+                .header("Content-Type", "text/plain");
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/cli/src/main/java/org/apache/ignite/cli/builtins/config/ConfigurationClient.java:84]: I've changed the Content-Type from applicaiton/json to text/plain because the HOCON configuration, in fact, is represented as a text (string) but not as a json object.

File: modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/ClusterConfigurationController.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    /**
+     * Updates cluster configuration in HOCON format.
+     *
+     * @param updatedConfiguration the cluster configuration to update.
+     */
+    @Operation(operationId = "updateClusterConfiguration")
+    @ApiResponse(responseCode = "200", description = "Configuration updated")
+    @Consumes(MediaType.TEXT_PLAIN)
+    @Patch
+    public void updateConfiguration(@Body String updatedConfiguration) throws Throwable {
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/ClusterConfigurationController.java:91]: Good point, will try.

File: modules/rest-api/src/main/java/org/apache/ignite/internal/rest/api/RestFactory.java [hunk 7]
```diff
@@ -15,8 +15,10 @@
...
 
+package org.apache.ignite.internal.rest.api;
+
 /**
- * This package contains classes that represent inbound requests, outbound responses and main REST handler.
+ * Factory that produces all beans that is necessary for the controller class.
  */
-
-package org.apache.ignite.internal.rest.netty;
+public interface RestFactory {
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/rest-api/src/main/java/org/apache/ignite/internal/rest/api/RestFactory.java:23]: This interface represents the factory that is needed for a controller, that is defined in a module. RestComponent takes the list of this factories and puts to the micronaut context.

File: modules/rest/pom.xml [hunk 8]
```diff
@@ -45,52 +45,41 @@
...
         </dependency>
 
         <dependency>
             <groupId>org.apache.ignite</groupId>
-            <artifactId>ignite-network</artifactId>
+            <artifactId>ignite-api</artifactId>
         </dependency>
 
         <!-- 3rd party dependencies -->
+        <!-- micronaut dependencies, use dependencyManagement when ignite-cli is migrated to micronaut 3 -->
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/rest/pom.xml:57]: I am going to apply this change in the current pull request. In case I face some problem I'll create one.

File: parent/pom.xml (hunk 2 above)
Comment (PakhomovAlexander) [reply parent/pom.xml:966]: Yes, you are right. I'll add a comment.

File: modules/cli/src/main/java/org/apache/ignite/cli/builtins/config/ConfigurationClient.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/cli/src/main/java/org/apache/ignite/cli/builtins/config/ConfigurationClient.java:84]: Makes sense, can you please leave a comment?

File: modules/rest-client/pom.xml (hunk 3 above)
Comment (PakhomovAlexander) [reply modules/rest-client/pom.xml:37]: Actually, this is a valuable note. 

This module (just module, not code) and pom are not generated. The plugin section in this pom configures the source code generation. This is done for integration tests for the generated client in the runner module.  

From one side, the client should be generated where it is really needed. In our case, it is inside the cli module. But the Open API spec is generated by the rest module and I think that there should be an integration test for the client that is generated from this Open API spec. And these checks have to be done by the server.

I can add a comment to the pom or drop this module and create a ticket for integration tests that will generate the client somewhere inside and not include it as a module.

File: modules/cluster-management/pom.xml [hunk 9]
```diff
@@ -49,17 +49,26 @@
...
         <dependency>
             <groupId>org.apache.ignite</groupId>
-            <artifactId>ignite-vault</artifactId>
+            <artifactId>ignite-rest-api</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>io.micronaut</groupId>
+            <artifactId>micronaut-http</artifactId>
+            <version>${micronaut.version}</version>
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/cluster-management/pom.xml:63]: As for micronaut-bom, it might not pass vulnerabilities checks https://mvnrepository.com/artifact/io.micronaut/micronaut-bom/3.5.0. 

One more thing is that it declares a huge set of dependencies that are not only micronaut-related. That could be a problem in dependency management.
https://repo1.maven.org/maven2/io/micronaut/micronaut-bom/3.5.0/micronaut-bom-3.5.0.pom

File: modules/rest-client/pom.xml (hunk 3 above)
Comment (sashapolo) [reply modules/rest-client/pom.xml:37]: > drop this module and create a ticket for integration tests that will generate the client somewhere inside and not include it as a module.

That would be very much preferable, thank you

File: modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/ClusterConfigurationController.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+     * Returns cluster configuration in HOCON format.
+     *
+     * @return the whole cluster configuration in HOCON format.
+     */
+    @Operation(operationId = "getClusterConfiguration")
+    @ApiResponse(
+            responseCode = "200",
+            content = @Content(mediaType = MediaType.TEXT_PLAIN, schema = @Schema(type = "string")),
+            description = "Get cluster configuration")
+    @Produces(MediaType.TEXT_PLAIN)
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/ClusterConfigurationController.java:59]: I'll add a comment.

File: modules/cluster-management/pom.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/cluster-management/pom.xml:63]: > As for micronaut-bom, it might not pass vulnerabilities checks

Why? There should be no difference between using a bom and declaring dependencies directly

> That could be a problem in dependency management.

Can you provide an example please?

[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/cluster-management/pom.xml:63]: > Why? There should be no difference between using a bom and declaring dependencies directly

Yes, you are right, there is no difference.

> Can you provide an example please?

Micronaut-bom includes a lot of other libraries' version definitions, for example, neo4j.version, h2.version, and many more. So, Ignite 3 will have them included to dependency management too.  Maybe that is not a problem but if Iginite 3 would like to support neo4j with a version that is different from micronaut-bom then we might face an issue.

[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/cluster-management/pom.xml:63]: AFAIK, if you declare a dependency directly in the dependency management section, it will override any artifacts declared in imported boms

File: modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/ClusterConfigurationController.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    @Get
+    public String getConfiguration() {
```
Comment (sashapolo) [inline modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/ClusterConfigurationController.java:62]: missing `@Override` here and on other methods

File: modules/configuration/src/test/java/org/apache/ignite/internal/rest/configuration/ConfigurationControllerBaseTest.java [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * The base test for configuration controllers.
+ */
+@MicronautTest
+public abstract class ConfigurationControllerBaseTest {
+    @Inject
+    EmbeddedServer server;
```
Comment (sashapolo) [inline modules/configuration/src/test/java/org/apache/ignite/internal/rest/configuration/ConfigurationControllerBaseTest.java:48]: Same question: can these fields be `private`?

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java [hunk 13]
```diff
@@ -17,151 +17,147 @@
...
+import org.apache.ignite.internal.rest.api.RestFactory;
+import org.apache.ignite.internal.rest.configuration.ClusterConfigurationController;
+import org.apache.ignite.internal.rest.configuration.NodeConfigurationController;
 import org.apache.ignite.lang.IgniteInternalException;
 import org.apache.ignite.lang.IgniteLogger;
-import org.apache.ignite.network.NettyBootstrapFactory;
 
 /**
- * Rest component is responsible for starting REST endpoints.
+ * Rest module is responsible for starting a REST endpoints for accessing and managing configuration.
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java:43]:
```suggestion
 * Rest module is responsible for starting REST endpoints for accessing and managing configuration.
```

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java [hunk 14]
```diff
@@ -17,151 +17,147 @@
...
-
-    private final SimpleRouter router = new SimpleRouter();
-
-    /** Netty channel. */
-    private volatile Channel channel;
+    /** Factories that produce beans needed for REST controllers. */
+    private final List<RestFactory> restFactories;
+    private final RestConfiguration restConfiguration;
+    /** Server host. */
+    private final String host = "localhost";
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java:61]: Why is this not a static constant?

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java [hunk 15]
```diff
@@ -17,151 +17,147 @@
...
-                ch = bindRes.channel();
+            try {
+                context = buildMicronautContext(portCandidate).start();
                 port = portCandidate;
+                log.info("REST protocol started successfully");
                 break;
-            } else if (!(bindRes.cause() instanceof BindException)) {
-                throw new RuntimeException(bindRes.cause());
+            } catch (ApplicationStartupException e) {
+                log.error("Got exception " + e.getCause() + " during node start on port " + portCandidate + " , trying again");
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java:90]: That's a bit strange. Are we sure that if we get this exception then the only reason for it is that the target port is busy?

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java [hunk 16]
```diff
@@ -17,151 +17,147 @@
...
-
-    /** Netty channel. */
-    private volatile Channel channel;
+    /** Factories that produce beans needed for REST controllers. */
+    private final List<RestFactory> restFactories;
+    private final RestConfiguration restConfiguration;
+    /** Server host. */
+    private final String host = "localhost";
+    /** Micronaut application context. */
+    private ApplicationContext context;
```
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java:63]: Actually, `port` can be left as-is if you carefully guard it with `context` accesses

File: modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/rest/exception/handler/IgniteInternalExceptionHandler.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+ */
+@Produces
+@Singleton
+@Requires(classes = {IgniteInternalException.class, ExceptionHandler.class})
+public class IgniteInternalExceptionHandler implements ExceptionHandler<IgniteInternalException, HttpResponse<ErrorResult>> {
+
+    @Override
+    public HttpResponse<ErrorResult> handle(HttpRequest request, IgniteInternalException exception) {
+        final ErrorResult errorResult = new ErrorResult("ALREADY_INITIALIZED", exception.getMessage());
+        return HttpResponseFactory.INSTANCE.status(HttpStatus.CONFLICT).body(errorResult);
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/rest/exception/handler/IgniteInternalExceptionHandler.java:42]: Since I've named it INITIALIZATION_EXCEPTION now this question is out of date.

File: modules/rest/pom.xml [hunk 18]
```diff
@@ -151,7 +128,21 @@
...
+                            <artifactId>micronaut-inject-java</artifactId>
+                            <version>${micronaut.version}</version>
+                        </path>
+                        <path>
+                            <groupId>io.micronaut.openapi</groupId>
+                            <artifactId>micronaut-openapi</artifactId>
+                            <version>${maven.micronaut.openapi.plugin.version}</version>
+                        </path>
                     </annotationProcessorPaths>
+                    <fork>true</fork>
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/rest/pom.xml:142]: Runnin the compiler in separate thread but I'll remove it in order not to confuse engineers.

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java (hunk 15 above)
Comment (PakhomovAlexander) [reply modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java:90]: Yes, this is strange. I'll rewrite it.

File: modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/ClusterConfigurationController.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/ClusterConfigurationController.java:60]: `application/hocon`?

File: modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/exception/handler/ConfigPathUnrecognizedExceptionHandler.java [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+ */
+@Produces(MediaType.APPLICATION_JSON)
+@Singleton
+@Requires(classes = {ConfigPathUnrecognizedException.class, ExceptionHandler.class})
+public class ConfigPathUnrecognizedExceptionHandler implements
+        ExceptionHandler<ConfigPathUnrecognizedException, HttpResponse<ErrorResult>> {
+
+    @Override
+    public HttpResponse<ErrorResult> handle(HttpRequest request, ConfigPathUnrecognizedException exception) {
+        final ErrorResult errorResult = new ErrorResult("CONFIG_PATH_UNRECOGNIZED", exception.getMessage());
```
Comment (rpuch) [inline modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/exception/handler/ConfigPathUnrecognizedExceptionHandler.java]: `final` seems redundant here as the method is just 2 lines

File: modules/configuration/src/test/java/org/apache/ignite/internal/rest/configuration/ClusterConfigurationControllerTest.java [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Inject
+    @Client("/management/v1/configuration/cluster/")
+    HttpClient client;
+
+    @Override
+    HttpClient client() {
+        return client;
+    }
+
```
Comment (rpuch) [inline modules/configuration/src/test/java/org/apache/ignite/internal/rest/configuration/ClusterConfigurationControllerTest.java:45]: An excessive blank line?

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java [hunk 21]
```diff
@@ -17,151 +17,147 @@
...
+    private void setFactories(Micronaut micronaut) {
+        for (var factory : restFactories) {
+            micronaut.singletons(factory);
         }
+    }
 
-        return ch.closeFuture();
+    private int mapServerStartupException(ServerStartupException exception) {
+        if (exception.getCause() instanceof BindException) {
+            return -1;
```
Comment (rpuch) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java:136]: What do these 1 and -1 mean in this case?

File: modules/cluster-management/src/integrationTest/java/org/apache/ignite/internal/cluster/management/rest/ItClusterManagementControllerTest.java [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    @Factory
+    @Bean
+    @Replaces(ClusterManagementRestFactory.class)
+    public ClusterManagementRestFactory clusterManagementRestFactory() {
+        return new ClusterManagementRestFactory(clusterService);
+    }
+
+    @NotNull
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/cluster-management/src/integrationTest/java/org/apache/ignite/internal/cluster/management/rest/ItClusterManagementControllerTest.java:173]: > It's not a JSON, but it's not just any plain text. Could we use a specific content type here? For example, here https://www.w3schools.io/file/hocon-introduction/ they suggest application/hocon; it does not seem to be registered with IANA, but it looks logical.

Yes, you are right, that sounds logical. But client generators get confused when you set content type that is not registered with IANA. I tried this.

File: modules/cli/src/integrationTest/java/org/apache/ignite/cli/ItConfigCommandTest.java [hunk 23]
```diff
@@ -113,9 +109,7 @@ public void setAndGetWithManualHost() {
 
         assertEquals(0, exitCode);
 
-        DocumentContext document = JsonPath.parse(removeTrailingQuotes(unescapeQuotes(out.toString(UTF_8))));
-
-        assertEquals(1, document.read("$.network.shutdownQuietPeriod", Integer.class));
+        assertTrue(out.toString(UTF_8).contains("\"shutdownQuietPeriod\" : 1"));
```
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/cli/src/integrationTest/java/org/apache/ignite/cli/ItConfigCommandTest.java:112]: > Is this caused by changing the content type from application/json to text/plain?  
Yes

File: modules/cli/src/integrationTest/java/org/apache/ignite/cli/ItConfigCommandTest.java [hunk 24]
```diff
@@ -169,11 +166,9 @@ public void partialGet() {
 
         assertEquals(0, exitCode);
 
-        JSONObject outResult = (JSONObject) JSONValue.parse(removeTrailingQuotes(unescapeQuotes(out.toString(UTF_8))));
-
-        assertTrue(outResult.containsKey("inbound"));
+        assertTrue(out.toString(UTF_8).contains("\"inbound\""));
```
Comment (sashapolo) [inline modules/cli/src/integrationTest/java/org/apache/ignite/cli/ItConfigCommandTest.java:169]: It's a bit strange that you decided to use `assertThat` above, but not here)

File: modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/rest/exception/handler/IgniteInternalExceptionHandler.java [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+ */
+@Produces
+@Singleton
+@Requires(classes = {IgniteInternalException.class, ExceptionHandler.class})
+public class IgniteInternalExceptionHandler implements ExceptionHandler<IgniteInternalException, HttpResponse<ErrorResult>> {
+
+    @Override
+    public HttpResponse<ErrorResult> handle(HttpRequest request, IgniteInternalException exception) {
+        ErrorResult errorResult = new ErrorResult("INITIALIZATION_EXCEPTION", exception.getMessage());
+        return HttpResponse.badRequest().body(errorResult);
```
Comment (sashapolo) [inline modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/rest/exception/handler/IgniteInternalExceptionHandler.java:40]: Are you sure it should be a "bad request` errror?

File: modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/exception/handler/IgniteExceptionHandler.java [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+ */
+@Produces
+@Singleton
+@Requires(classes = {IgniteException.class, ExceptionHandler.class})
+public class IgniteExceptionHandler implements ExceptionHandler<IgniteException, HttpResponse<ErrorResult>> {
+
+    @Override
+    public HttpResponse<ErrorResult> handle(HttpRequest request, IgniteException exception) {
+        ErrorResult errorResult = new ErrorResult("APPLICATION_EXCEPTION", exception.getMessage());
+        return HttpResponse.badRequest().body(errorResult);
```
Comment (sashapolo) [inline modules/configuration/src/main/java/org/apache/ignite/internal/rest/configuration/exception/handler/IgniteExceptionHandler.java:41]: Same here, I don't think it should be a "bad request"

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java [hunk 27]
```diff
@@ -17,151 +17,161 @@
...
-                .childHandler(new RestApiInitializer(router));
-
         for (int portCandidate = desiredPort; portCandidate <= desiredPort + portRange; portCandidate++) {
-            ChannelFuture bindRes = bootstrap.bind(portCandidate).awaitUninterruptibly();
-
-            if (bindRes.isSuccess()) {
-                ch = bindRes.channel();
+            try {
+                context = buildMicronautContext(portCandidate).start();
                 port = portCandidate;
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java:90]: unfortunately this is not correct. If you want to avoid making `port` variable `volatile` and guard it with `context` accesses, you must always write to the `port` variable **before** writing to the `context` variable. Same applies to reading, but in reverse order

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java [hunk 28]
```diff
@@ -17,151 +17,161 @@
...
+                log.info("REST protocol started successfully");
+                return;
+            } catch (ApplicationStartupException e) {
+                if (isBindExceptionInside(e)) {
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java:94]: I would suggest to introduce a `findBindException` method that will return an exception or `null`. `e.getCause().getCause().getCause()` looks really weird

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java [hunk 29]
```diff
@@ -17,151 +17,161 @@
...
+                    log.error("Got exception " + e.getCause().getCause().getCause() + " during node start on port "
+                            + portCandidate + " , trying again");
+                    continue;
+                }
                 break;
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java:99]: Looks like in case of `e` not being a `BindException` you will lose the exception and will not log it. I think it makes sense to simply re-throw it

File: parent/pom.xml [hunk 30]
```diff
@@ -473,6 +475,55 @@
                 <version>${micronaut.version}</version>
             </dependency>
 
+            <dependency>
+                <groupId>io.micronaut</groupId>
```
Comment (sashapolo) [inline parent/pom.xml:479]: So, what about using a `bom`?

File: modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/rest/exception/handler/IgniteInternalExceptionHandler.java (hunk 25 above)
Comment (PakhomovAlexander) [reply modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/rest/exception/handler/IgniteInternalExceptionHandler.java:40]: Not it is not obvious what exactly happend on IgniteInternalException. `IgniteInternalException` looks like 500 and I think it is better to mark it as an internal server error.

Comment (PakhomovAlexander) [reply modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/rest/exception/handler/IgniteInternalExceptionHandler.java:40]: In future we have to introduce specific exceptions for situations like: already initialized, initializing, etc.

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java (hunk 27 above)
Comment (PakhomovAlexander) [reply modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java:90]: Thanks, you are absolutelly right.

File: modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/rest/exception/handler/IgniteInternalExceptionHandler.java (hunk 25 above)
Comment (sashapolo) [reply modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/rest/exception/handler/IgniteInternalExceptionHandler.java:40]: > IgniteInternalException looks like 500 and I think it is better to mark it as an internal server error.

Exactly

File: parent/pom.xml (hunk 30 above)
Comment (PakhomovAlexander) [reply parent/pom.xml:479]: micronaut-bom downgrades <junit5.version>5.7.2</junit5.version> , but ignite 3 uses 5.8.1 version. It actually breaks the compilation. Are we ok to downgrade the junit version?

File: modules/cluster-management/pom.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (PakhomovAlexander) [reply modules/cluster-management/pom.xml]: What about two boms have different versions?

File: parent/pom.xml (hunk 30 above)
Comment (sashapolo) [reply parent/pom.xml:479]: I don't understand how this can happen. Shouldn't our explicit dependency override what is stated in the bom?

Comment (PakhomovAlexander) [reply parent/pom.xml:479]: junit-bom does not override this

**Review by sashapolo — APPROVED** [review]

**Review by ptupitsyn — APPROVED** [review]
